// @accordproject/concerto-linter-default-ruleset/dist/pascal-case-declarations.js
"use strict";
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const spectral_functions_1 = require("@stoplight/spectral-functions");
/**
 * Rule: Pascal Case Declarations
 * ------------------------------
 * Ensures that declaration names (scalar, enum, concept, asset, participant, transaction, event)
 * follow PascalCase naming convention (e.g., 'MyDeclaration').
 * This promotes consistency and readability across model declarations.
 */
exports.default = {
    description: 'Declaration names (scalar, enum, concept, asset, participant, transaction, event) should be PascalCase.',
    given: '$.models[*].declarations[*].name',
    message: 'Declaration \'{{value}}\' should be PascalCase (e.g. \'MyDeclaration\')',
    severity: 0, // 0 = error, 1 = warning, 2 = info, 3 = hint
    then: {
        function: spectral_functions_1.casing,
        functionOptions: {
            type: 'pascal',
        },
    },
};
