// @accordproject/concerto-analysis/dist/compare.js
"use strict";
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Compare = void 0;
const concerto_core_1 = require("@accordproject/concerto-core");
const compare_config_1 = require("./compare-config");
class Compare {
    constructor(config) {
        this.config = Object.assign({}, compare_config_1.defaultCompareConfig, config);
    }
    compare(a, b) {
        if (!a.getModelManager().isStrict()) {
            throw new Error(`model file "${a.getNamespace()}" does not have strict versioned namespaces`);
        }
        else if (!b.getModelManager().isStrict()) {
            throw new Error(`model file "${b.getNamespace()}" does not have strict versioned namespaces`);
        }
        const comparerFactories = this.config.comparerFactories;
        const findings = [];
        const comparers = comparerFactories.map(comparerFactory => comparerFactory({
            report: finding => findings.push(finding),
        }));
        this.compareModelFiles(comparers, a, b);
        return this.buildResults(findings);
    }
    getAddedProperties(a, b) {
        return b.filter(bItem => !a.some(aItem => bItem.getName() === aItem.getName()));
    }
    getMatchingProperties(a, b) {
        return a.map(aItem => [aItem, b.find(bItem => aItem.getName() === bItem.getName())]).filter(([, b]) => !!b);
    }
    getRemovedProperties(a, b) {
        return a.filter(aItem => !b.some(bItem => aItem.getName() === bItem.getName()));
    }
    compareProperty(comparers, a, b) {
        comparers.forEach(comparer => { var _a; return (_a = comparer.compareProperty) === null || _a === void 0 ? void 0 : _a.call(comparer, a, b); });
    }
    compareProperties(comparers, a, b) {
        const added = this.getAddedProperties(a, b);
        const matching = this.getMatchingProperties(a, b);
        const removed = this.getRemovedProperties(a, b);
        added.forEach(b => comparers.forEach(comparer => { var _a; return (_a = comparer.compareProperty) === null || _a === void 0 ? void 0 : _a.call(comparer, undefined, b); }));
        matching.forEach(([a, b]) => this.compareProperty(comparers, a, b));
        removed.forEach(a => comparers.forEach(comparer => { var _a; return (_a = comparer.compareProperty) === null || _a === void 0 ? void 0 : _a.call(comparer, a, undefined); }));
    }
    getAddedMapDeclarations(a, b) {
        return b.filter(bItem => !a.some(aItem => bItem.getName() === aItem.getName()));
    }
    getAddedScalarDeclarations(a, b) {
        return b.filter(bItem => !a.some(aItem => bItem.getName() === aItem.getName()));
    }
    getMatchingMapDeclarations(a, b) {
        return a.map(aItem => [aItem, b.find(bItem => aItem.getName() === bItem.getName())]).filter(([, b]) => !!b);
    }
    getMatchingScalarDeclarations(a, b) {
        return a.map(aItem => [aItem, b.find(bItem => aItem.getName() === bItem.getName())]).filter(([, b]) => !!b);
    }
    getRemovedMapDeclarations(a, b) {
        return a.filter(aItem => !b.some(bItem => aItem.getName() === bItem.getName()));
    }
    getRemovedScalarDeclarations(a, b) {
        return a.filter(aItem => !b.some(bItem => aItem.getName() === bItem.getName()));
    }
    getAddedClassDeclarations(a, b) {
        return b.filter(bItem => !a.some(aItem => bItem.getName() === aItem.getName()));
    }
    getMatchingClassDeclarations(a, b) {
        return a.map(aItem => [aItem, b.find(bItem => aItem.getName() === bItem.getName())]).filter(([, b]) => !!b);
    }
    getRemovedClassDeclarations(a, b) {
        return a.filter(aItem => !b.some(bItem => aItem.getName() === bItem.getName()));
    }
    compareMapDeclarations(comparers, a, b) {
        const added = this.getAddedMapDeclarations(a, b);
        const matching = this.getMatchingMapDeclarations(a, b);
        const removed = this.getRemovedMapDeclarations(a, b);
        added.forEach(b => comparers.forEach(comparer => { var _a; return (_a = comparer.compareMapDeclaration) === null || _a === void 0 ? void 0 : _a.call(comparer, undefined, b); }));
        matching.forEach(([a, b]) => comparers.forEach(comparer => { var _a; return (_a = comparer.compareMapDeclaration) === null || _a === void 0 ? void 0 : _a.call(comparer, a, b); }));
        removed.forEach(a => comparers.forEach(comparer => { var _a; return (_a = comparer.compareMapDeclaration) === null || _a === void 0 ? void 0 : _a.call(comparer, a, undefined); }));
    }
    compareScalarDeclarations(comparers, a, b) {
        const added = this.getAddedScalarDeclarations(a, b);
        const matching = this.getMatchingScalarDeclarations(a, b);
        const removed = this.getRemovedScalarDeclarations(a, b);
        added.forEach(b => comparers.forEach(comparer => { var _a; return (_a = comparer.compareScalarDeclaration) === null || _a === void 0 ? void 0 : _a.call(comparer, undefined, b); }));
        matching.forEach(([a, b]) => comparers.forEach(comparer => { var _a; return (_a = comparer.compareScalarDeclaration) === null || _a === void 0 ? void 0 : _a.call(comparer, a, b); }));
        removed.forEach(a => comparers.forEach(comparer => { var _a; return (_a = comparer.compareScalarDeclaration) === null || _a === void 0 ? void 0 : _a.call(comparer, a, undefined); }));
    }
    compareClassDeclaration(comparers, a, b) {
        comparers.forEach(comparer => { var _a; return (_a = comparer.compareClassDeclaration) === null || _a === void 0 ? void 0 : _a.call(comparer, a, b); });
        // MapDeclarations do not contain properties, nothing to compare.
        if (a instanceof concerto_core_1.MapDeclaration || b instanceof concerto_core_1.MapDeclaration) {
            return;
        }
        // ScalarDeclarations do not contain properties, nothing to compare.
        if (a instanceof concerto_core_1.ScalarDeclaration || b instanceof concerto_core_1.ScalarDeclaration) {
            return;
        }
        this.compareProperties(comparers, a.getOwnProperties(), b.getOwnProperties());
    }
    compareClassDeclarations(comparers, a, b) {
        const added = this.getAddedClassDeclarations(a, b);
        const matching = this.getMatchingClassDeclarations(a, b);
        const removed = this.getRemovedClassDeclarations(a, b);
        added.forEach(b => comparers.forEach(comparer => { var _a; return (_a = comparer.compareClassDeclaration) === null || _a === void 0 ? void 0 : _a.call(comparer, undefined, b); }));
        matching.forEach(([a, b]) => this.compareClassDeclaration(comparers, a, b));
        removed.forEach(a => comparers.forEach(comparer => { var _a; return (_a = comparer.compareClassDeclaration) === null || _a === void 0 ? void 0 : _a.call(comparer, a, undefined); }));
    }
    compareModelFiles(comparers, a, b) {
        comparers.forEach(comparer => { var _a; return (_a = comparer.compareModelFiles) === null || _a === void 0 ? void 0 : _a.call(comparer, a, b); });
        this.compareClassDeclarations(comparers, a.getAllDeclarations(), b.getAllDeclarations());
        this.compareMapDeclarations(comparers, a.getMapDeclarations(), b.getMapDeclarations());
        this.compareScalarDeclarations(comparers, a.getScalarDeclarations(), b.getScalarDeclarations());
    }
    buildResults(findings) {
        const results = {
            findings: [],
            result: compare_config_1.CompareResult.NONE,
        };
        findings.forEach(finding => {
            var _a;
            const result = (_a = this.config.rules[finding.key]) !== null && _a !== void 0 ? _a : compare_config_1.CompareResult.NONE;
            results.findings.push(Object.assign(Object.assign({}, finding), { result }));
            if (result > results.result) {
                results.result = result;
            }
        });
        return results;
    }
}
exports.Compare = Compare;
